{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Global variables\n",
    "ROOT = \"/users/kcochran/projects/domain_adaptation_nosexchr/\"\n",
    "\n",
    "tfs = [\"CTCF\", \"CEBPA\", \"Hnf4a\", \"RXRA\"]\n",
    "tfs_latex_names = [\"CTCF\", \"CEBPA\", \"HNF4A\", \"RXRA\"]\n",
    "\n",
    "all_trainspecies = [\"mm10\", \"hg38\"]\n",
    "model_names_dict = {\"mm10\" : \"Mouse\", \"hg38\" : \"Human\"}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Predictions and Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== CTCF, mm10-trained ===\n",
      "=== CEBPA, mm10-trained ===\n",
      "=== Hnf4a, mm10-trained ===\n",
      "=== RXRA, mm10-trained ===\n",
      "=== CTCF, hg38-trained ===\n",
      "=== CEBPA, hg38-trained ===\n",
      "=== Hnf4a, hg38-trained ===\n",
      "=== RXRA, hg38-trained ===\n"
     ]
    }
   ],
   "source": [
    "def get_preds_file(tf, train_species, test_species):\n",
    "    # this file is created by the 0_generate_predictions notebook\n",
    "    return ROOT + \"/model_out/\" + tf + \"_\" + train_species + \"-trained_\" + test_species + \"-test.preds.npy\"\n",
    "\n",
    "\n",
    "def load_average_test_set_preds(test_species):\n",
    "    # takes a while to run.\n",
    "    preds_dict = defaultdict(lambda : dict())\n",
    "\n",
    "    # loop over mouse-trained, human-trained models, and DA mouse-trained models\n",
    "    for train_species in all_trainspecies:\n",
    "        for tf in tfs:\n",
    "            print(\"=== \" + tf + \", \" + train_species + \"-trained ===\")\n",
    "\n",
    "            # load predictions for all 5 independent model runs\n",
    "            preds_file = get_preds_file(tf, train_species, test_species)\n",
    "            try:\n",
    "                preds_dict[train_species][tf] = np.mean(np.load(preds_file), axis = 1)\n",
    "            except:\n",
    "                print(\"Could not load preds file:\", preds_file)\n",
    "            \n",
    "    return preds_dict\n",
    "\n",
    "\n",
    "avg_preds_human_test = load_average_test_set_preds(\"hg38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_test_bed_file(tf, species):\n",
    "    # should be in BED file format\n",
    "    # this file is specific to each tf -- the last column\n",
    "    # should contain the binding label for each window\n",
    "    return(ROOT + \"data/\" + species + \"/\" + tf + \"/chr2.bed\")\n",
    "\n",
    "\n",
    "def get_test_labels(tf, species = \"hg38\"):\n",
    "    # This function reads in the test-data bed file \n",
    "    # for a given species and TF and returns the binding labels\n",
    "    # for each example in that file.\n",
    "    labels_file = get_test_bed_file(tf, species)\n",
    "    with open(labels_file) as f:\n",
    "        return np.array([int(line.split()[-1]) for line in f])\n",
    "        \n",
    "\n",
    "def get_all_test_labels_and_indices(preds_dict, species = \"hg38\"):\n",
    "    labels = dict()\n",
    "    bound_indices = dict()\n",
    "    for tf in tfs:\n",
    "        # load binding labels from bed file\n",
    "        labels_for_tf = get_test_labels(tf)\n",
    "        # assuming the mm10 entry exists, species doesn't matter here\n",
    "        # shape of preds array should be the same in all cases\n",
    "        len_to_truncate_by = preds_dict[\"mm10\"][tf].shape[0]\n",
    "        # truncate binding labels to be multiple of batch size, like preds\n",
    "        labels[tf] = labels_for_tf[:len_to_truncate_by]\n",
    "        # get indices where label is bound\n",
    "        bound_indices[tf] = np.nonzero(labels[tf])[0]\n",
    "    return labels, bound_indices\n",
    "\n",
    "labels, bound_indices = get_all_test_labels_and_indices(avg_preds_human_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find FPs, FNs, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTCF mm10\n",
      "CTCF hg38\n",
      "CEBPA mm10\n",
      "CEBPA hg38\n",
      "Hnf4a mm10\n",
      "Hnf4a hg38\n",
      "RXRA mm10\n",
      "RXRA hg38\n"
     ]
    }
   ],
   "source": [
    "def get_fp_fn_indices(preds_dict, bound_indices):\n",
    "    #tp_indices = defaultdict(lambda : defaultdict(lambda : []))  # not needed for now\n",
    "    fp_indices = defaultdict(lambda : defaultdict(lambda : []))\n",
    "    fn_indices = defaultdict(lambda : defaultdict(lambda : []))\n",
    "\n",
    "    for tf in tfs:\n",
    "        for train_species in all_trainspecies:\n",
    "            print(tf, train_species)\n",
    "            # using a threshold of 0.5 here to decide what is predicted as bound\n",
    "            predicted_as_bound_indices = np.nonzero(preds_dict[train_species][tf] > 0.5)[0]\n",
    "\n",
    "            for index in predicted_as_bound_indices:\n",
    "                if not index in bound_indices[tf]:\n",
    "                    fp_indices[tf][train_species].append(index)\n",
    "                #else:\n",
    "                #    tp_indices[tf][train_species].append(index)\n",
    "            for index in bound_indices[tf]:\n",
    "                if index not in predicted_as_bound_indices:\n",
    "                    fn_indices[tf][train_species].append(index)\n",
    "            \n",
    "    return fp_indices, fn_indices\n",
    "\n",
    "\n",
    "def get_site_indices(preds_dict, bound_indices):\n",
    "    # This function uses the predictions and binding labels for\n",
    "    # all sites in the test set to determine if sites are\n",
    "    # both-model false positives, false negatives, mouse-model FPs, etc.\n",
    "    \n",
    "    # Here we use a definition of \"differentially predicted site\"\n",
    "    # to find mouse-model FPs or FNs that requires the difference\n",
    "    # between the mouse-model prediction and the human-model prediction\n",
    "    # to be at least 0.5\n",
    "    fp_indices, fn_indices = get_fp_fn_indices(preds_dict, bound_indices)\n",
    "    \n",
    "    site_indices = defaultdict(lambda : dict())\n",
    "\n",
    "    for tf in tfs:\n",
    "        site_indices[tf][\"bothFP\"] = set(fp_indices[tf][\"mm10\"]).intersection(set(fp_indices[tf][\"hg38\"]))\n",
    "        site_indices[tf][\"bothFN\"] = set(fn_indices[tf][\"mm10\"]).intersection(set(fn_indices[tf][\"hg38\"]))\n",
    "\n",
    "        diff_pred_mm10_overpred = set(np.nonzero(preds_dict[\"mm10\"][tf] - preds_dict[\"hg38\"][tf] > 0.5)[0])\n",
    "        diff_pred_mm10_underpred = set(np.nonzero(preds_dict[\"hg38\"][tf] - preds_dict[\"mm10\"][tf] > 0.5)[0])\n",
    "\n",
    "        site_indices[tf][\"mFP\"] = set(fp_indices[tf][\"mm10\"]).intersection(diff_pred_mm10_overpred)\n",
    "        site_indices[tf][\"mFN\"] = set(fn_indices[tf][\"mm10\"]).intersection(diff_pred_mm10_underpred)\n",
    "        \n",
    "    return site_indices\n",
    "\n",
    "\n",
    "site_subset_indices = get_site_indices(avg_preds_human_test, bound_indices)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Repeat Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA\n",
      "LINE\n",
      "Low_complexity\n",
      "LTR\n",
      "Simple_repeat\n",
      "SINE\n",
      "Unknown\n"
     ]
    }
   ],
   "source": [
    "REPEAT_TYPES = [\"DNA\", \"LINE\", \"Low_complexity\", \"LTR\", \"Simple_repeat\", \"SINE\", \"Unknown\"]\n",
    "# Removed due to < 500 instances in the test set: [\"RC\", \"Retroposon\", \"RNA\", \"rRNA\", \"Satellite\", \"scRNA\", \"srpRNA\", \"tRNA\"]\n",
    "\n",
    "def get_rmsk_file():\n",
    "    # this file is downloaded by an earlier script\n",
    "    return ROOT + \"data/hg38/rmsk.bed\"\n",
    "\n",
    "\n",
    "def read_and_filter_rmsk_file(repeat_name, is_subfam = False, test_chrom = \"chr2\"):\n",
    "    # This function reads in the RepeatMasker bed file,\n",
    "    # filters for only rows listing annotations of one\n",
    "    # repeat type, and then returns only the start and \n",
    "    # end coordinate for each annotation.\n",
    "    \n",
    "    # We're assuming all the test set examples are on\n",
    "    # one chromosome, so we don't need the first column.\n",
    "    \n",
    "    # assuming bed format\n",
    "    filename = get_rmsk_file()\n",
    "    df = pd.read_csv(filename, sep = \"\\t\", usecols = [5, 6, 7, 10, 11], header = None)\n",
    "\n",
    "    if is_subfam:\n",
    "        df = df[df[10] == repeat_name]\n",
    "        df = df[df[5] == test_chrom]\n",
    "    else:\n",
    "        df = df[df[11] == repeat_name]\n",
    "        df = df[df[5] == test_chrom]\n",
    "\n",
    "    sorted_repeat_coords = sorted(list(zip(df[6], df[7])), key = lambda tup : tup[0])\n",
    "    return np.array(sorted_repeat_coords)\n",
    "\n",
    "\n",
    "def get_repeat_and_test_set_overlap(list_a, list_b):\n",
    "    # This function is similar to bedtools intersect,\n",
    "    # but returns a binary yes/no for overlap for each\n",
    "    # window in list_a.\n",
    "    \n",
    "    # Assumes everything's on the same chromosome\n",
    "    # Assumes inputs are lists of 2-ples: (start, stop) \n",
    "    \n",
    "    # output is list with len == len(list_a)\n",
    "    matches = []\n",
    "    b_index = 0\n",
    "    for a_item in list_a:\n",
    "        a_start, a_end = a_item\n",
    "        while True:\n",
    "            if b_index >= len(list_b):\n",
    "                matches.append(False)\n",
    "                break\n",
    "                \n",
    "            b_start, b_end = list_b[b_index]\n",
    "            # the -1 is because bed files are 1-indexed\n",
    "            if b_start > a_end - 1:  \n",
    "                matches.append(False)\n",
    "                break\n",
    "            elif b_end <= a_start:\n",
    "                b_index += 1\n",
    "            else:\n",
    "                matches.append(True)\n",
    "                break\n",
    "    assert len(matches) == len(list_a)\n",
    "    return np.array(matches)\n",
    "\n",
    "\n",
    "def get_test_bed_coords(species = \"hg38\"):\n",
    "    # This function loads in the bed file for the test set\n",
    "    # and keeps only the start and end coords for each entry.\n",
    "    # Here we assume the test set is 1 chromosome\n",
    "    \n",
    "    # later analysis will assume the coords are sorted,\n",
    "    # as in `sort -k1,1 -k2,2n $bed_file`\n",
    "    \n",
    "    # TF doesn't matter here because we're not using labels\n",
    "    test_bed = get_test_bed_file(tfs[0], species)\n",
    "    df = pd.read_csv(test_bed, sep = \"\\t\", usecols = [1, 2],\n",
    "                     header = None)\n",
    "    return df.to_numpy()\n",
    "\n",
    "\n",
    "def get_all_repeat_labels_and_indices():\n",
    "    all_windows_coords = get_test_bed_coords()\n",
    "    repeat_labels = dict()\n",
    "    repeat_indices = dict()\n",
    "\n",
    "    for repeat_type in REPEAT_TYPES:\n",
    "        print(repeat_type)\n",
    "        repeat_type_coords = read_and_filter_rmsk_file(repeat_type)\n",
    "        # filtering for repeat types with at least 500 instances\n",
    "        # in the test set, so we don't get incorrectly extreme results\n",
    "        assert len(repeat_type_coords) > 500, (repeat_type, len(repeat_type_coords))\n",
    "        repeat_labels[repeat_type] = get_repeat_and_test_set_overlap(all_windows_coords, repeat_type_coords)\n",
    "        repeat_indices[repeat_type] = set(np.nonzero(repeat_labels[repeat_type])[0])\n",
    "        \n",
    "    return repeat_labels, repeat_indices\n",
    "\n",
    "\n",
    "repeat_labels, repeat_indices = get_all_repeat_labels_and_indices()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_repeat_fracs_from_site_overlap(bound_indices, labels, site_subset_indices, repeat_indices):\n",
    "    # This function calculates what fraction of sites are overlapping a given repeat type,\n",
    "    # for various categories of sites. The output is returned in nested dictionaries, one\n",
    "    # for each TF, because the site categorizations are specific to the TF.\n",
    "    \n",
    "    repeat_fracs = defaultdict(lambda : dict())\n",
    "\n",
    "    for tf in tfs:\n",
    "        # Bound site repeat fraction = (# bound sites overlapping repeat) / (# bound sites)\n",
    "        num_bound_sites_with_repeat = len(set(bound_indices[tf]).intersection(repeat_indices))\n",
    "        repeat_fracs[tf][\"bound\"] = num_bound_sites_with_repeat / len(bound_indices[tf])\n",
    "        \n",
    "        # this arithmetic reverses binary-numeric labels (0 is now 1, 1 is now 0)\n",
    "        num_unbound_sites = sum(labels[tf] * -1 + 1)\n",
    "        \n",
    "        # Unbound site repeat fraction = (# unbound sites overlapping repeat) / (# unbound sites)\n",
    "        # where (# unbound sites overlapping repeat) = (# repeat-overlap sites in test set not in bound site set)\n",
    "        num_unbound_sites_with_repeat = len(repeat_indices.difference(set(bound_indices[tf])))\n",
    "        repeat_fracs[tf][\"unbound\"] = num_unbound_sites_with_repeat / num_unbound_sites\n",
    "\n",
    "        # for each of the specific categories of sites we're interested in...\n",
    "        # (e.g. \"false positives\", \"mouse-model false negatives\")\n",
    "        for site_type in site_subset_indices[tf].keys():\n",
    "            # calc total # of sites in this category\n",
    "            num_sites = len(site_subset_indices[tf][site_type])\n",
    "            if num_sites > 0:\n",
    "                # calc # of sites in this category that overlap the given repeat type\n",
    "                num_sites_with_repeat = len(site_subset_indices[tf][site_type].intersection(repeat_indices))\n",
    "                # finally, calc fraction of sites in this category that overlap the repeat type\n",
    "                repeat_fracs[tf][site_type] = num_sites_with_repeat / num_sites\n",
    "            else:\n",
    "                repeat_fracs[tf][site_type] = np.nan\n",
    "            \n",
    "    return repeat_fracs\n",
    "\n",
    "\n",
    "def generate_all_repeat_fracs_for_table(bound_indices, labels, site_subset_indices, repeat_indices):     \n",
    "    # This function creates the dictionary needed for the print_full_table() function below,\n",
    "    # where each key is a repeat type and the value is nested dictionaries of the fraction\n",
    "    # of sites overlapping that repeat type for a given TF (since sets of sites, such as\n",
    "    # \"bound site\" and \"mouse-model false positive\", are different for each TF).\n",
    "    repeat_fracs = dict()\n",
    "\n",
    "    for repeat_type in REPEAT_TYPES:\n",
    "        repeat_fracs[repeat_type] = calc_repeat_fracs_from_site_overlap(bound_indices, labels,\n",
    "                                                                        site_subset_indices,\n",
    "                                                                        repeat_indices[repeat_type])\n",
    "    return repeat_fracs\n",
    "\n",
    "\n",
    "\n",
    "all_repeat_fracs = generate_all_repeat_fracs_for_table(bound_indices, labels,\n",
    "                                                       site_subset_indices,\n",
    "                                                       repeat_indices)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{table*}{\n",
      "\\setlength{\\tabcolsep}{0.8em}\n",
      "\\centering \\begin{tabular}{@{}ccccccc@{}}\\toprule\n",
      "TF & Bound & FN (Both Models) & FN (Mouse Only) & Unbound & FP (Both Models) & FP (Mouse Only) \\\\\\midrule\n",
      "\\multicolumn{7}{c}{\\textbf{DNA}} \\\\\\midrule\n",
      "CTCF & 10.1\\% & 11.4\\% & 7.3\\% & 11.4\\% & 8.9\\% & \\textbf{9.0\\%} \\\\\n",
      "CEBPA & 12.3\\% & 10.4\\% & 8.3\\% & 11.3\\% & 13.0\\% & \\textbf{9.2\\%} \\\\\n",
      "HNF4A & 10.7\\% & 12.0\\% & 11.3\\% & 11.4\\% & 9.5\\% & \\textbf{9.0\\%} \\\\\n",
      "RXRA & 10.1\\% & 11.7\\% & 8.8\\% & 11.4\\% & 10.0\\% & \\textbf{9.4\\%} \\\\\\midrule\n",
      "\\multicolumn{7}{c}{\\textbf{LINE}} \\\\\\midrule\n",
      "CTCF & 18.3\\% & 22.5\\% & 21.3\\% & 37.6\\% & 17.8\\% & \\textbf{31.6\\%} \\\\\n",
      "CEBPA & 25.6\\% & 26.3\\% & 25.0\\% & 37.6\\% & 29.0\\% & \\textbf{32.3\\%} \\\\\n",
      "HNF4A & 21.0\\% & 25.3\\% & 26.3\\% & 37.6\\% & 21.5\\% & \\textbf{30.5\\%} \\\\\n",
      "RXRA & 21.0\\% & 27.9\\% & 22.1\\% & 37.8\\% & 22.1\\% & \\textbf{33.2\\%} \\\\\\midrule\n",
      "\\multicolumn{7}{c}{\\textbf{Low Complexity}} \\\\\\midrule\n",
      "CTCF & 2.5\\% & 1.0\\% & 2.6\\% & 1.9\\% & 4.0\\% & \\textbf{1.5\\%} \\\\\n",
      "CEBPA & 1.5\\% & 1.4\\% & 0.0\\% & 1.9\\% & 1.6\\% & \\textbf{1.4\\%} \\\\\n",
      "HNF4A & 2.0\\% & 1.1\\% & 1.5\\% & 1.9\\% & 2.4\\% & \\textbf{1.3\\%} \\\\\n",
      "RXRA & 2.1\\% & 1.3\\% & 2.0\\% & 1.9\\% & 2.2\\% & \\textbf{1.5\\%} \\\\\\midrule\n",
      "\\multicolumn{7}{c}{\\textbf{LTR}} \\\\\\midrule\n",
      "CTCF & 8.6\\% & 12.6\\% & 8.2\\% & 17.6\\% & 16.4\\% & \\textbf{15.5\\%} \\\\\n",
      "CEBPA & 12.8\\% & 12.5\\% & 41.7\\% & 17.6\\% & 19.4\\% & \\textbf{14.0\\%} \\\\\n",
      "HNF4A & 13.3\\% & 15.0\\% & 16.0\\% & 17.6\\% & 19.4\\% & \\textbf{12.6\\%} \\\\\n",
      "RXRA & 12.2\\% & 14.8\\% & 9.8\\% & 17.6\\% & 18.2\\% & \\textbf{11.5\\%} \\\\\\midrule\n",
      "\\multicolumn{7}{c}{\\textbf{Simple Repeat}} \\\\\\midrule\n",
      "CTCF & 13.4\\% & 10.1\\% & 10.9\\% & 11.5\\% & 15.8\\% & \\textbf{12.0\\%} \\\\\n",
      "CEBPA & 9.4\\% & 8.6\\% & 25.0\\% & 11.6\\% & 9.8\\% & \\textbf{11.9\\%} \\\\\n",
      "HNF4A & 12.3\\% & 11.4\\% & 9.1\\% & 11.5\\% & 11.6\\% & \\textbf{12.8\\%} \\\\\n",
      "RXRA & 11.5\\% & 9.4\\% & 13.3\\% & 11.5\\% & 11.1\\% & \\textbf{13.4\\%} \\\\\\midrule\n",
      "\\multicolumn{7}{c}{\\textbf{SINE}} \\\\\\midrule\n",
      "CTCF & 23.9\\% & 23.2\\% & 20.6\\% & 31.2\\% & 18.4\\% & \\textbf{81.1\\%} \\\\\n",
      "CEBPA & 30.9\\% & 22.6\\% & 41.7\\% & 31.2\\% & 35.2\\% & \\textbf{86.9\\%} \\\\\n",
      "HNF4A & 27.1\\% & 23.6\\% & 22.6\\% & 31.2\\% & 28.9\\% & \\textbf{95.5\\%} \\\\\n",
      "RXRA & 27.1\\% & 25.3\\% & 18.9\\% & 31.3\\% & 32.2\\% & \\textbf{97.6\\%} \\\\\\midrule\n",
      "\\multicolumn{7}{c}{\\textbf{Unknown}} \\\\\\midrule\n",
      "CTCF & 0.2\\% & 0.1\\% & 0.0\\% & 0.2\\% & 0.1\\% & \\textbf{0.0\\%} \\\\\n",
      "CEBPA & 0.3\\% & 0.5\\% & 0.0\\% & 0.1\\% & 0.2\\% & \\textbf{0.0\\%} \\\\\n",
      "HNF4A & 0.2\\% & 0.1\\% & 0.0\\% & 0.2\\% & 0.2\\% & \\textbf{0.0\\%} \\\\\n",
      "RXRA & 0.2\\% & 0.3\\% & 0.0\\% & 0.1\\% & 0.2\\% & \\textbf{0.0\\%} \\\\\\bottomrule\n",
      "\\end{tabular}}{}\n",
      "\\captionof{table}{Percent of windows overlapping various RepeatMasker-defined repeat elements, for different categories of genomic windows from the held-out test set. Only RepeatMasker repeat classes with at least 500 distinct annotations within the test set are shown. FPs: false positives. FNs: false negatives. Mouse Only: specific tomouse-trained models. See Methods for more details on site categorization.}\n",
      "\\end{table*}\n"
     ]
    }
   ],
   "source": [
    "def fix_repeat_name(repeat_name):\n",
    "    # If a repeat name has an underscore in it, this will\n",
    "    # mess up the latex formatting, so we replace the\n",
    "    # underscores with spaces and then capitalize the first\n",
    "    # letter of each word in the repeat name\n",
    "    if \"_\" in repeat_name:\n",
    "        return \" \".join(repeat_name.split(\"_\")).title()\n",
    "    return repeat_name\n",
    "    \n",
    "\n",
    "def print_full_table(all_repeat_fracs, header = None, row_order = None):\n",
    "    # This function prints a fully latex-formatted table, made up of\n",
    "    # one sub-table for each repeat type in the all_repeat_fracs dict.\n",
    "    # The columns of the table are given below in header, and the\n",
    "    # rows of the sub-tables are the TFs.\n",
    "    print(r'\\begin{table*}{')\n",
    "    print(r'\\setlength{\\tabcolsep}{0.8em}')\n",
    "    print(r'\\centering \\begin{tabular}{@{}ccccccc@{}}\\toprule')\n",
    "    \n",
    "    if header is None:\n",
    "        header = \"TF & Bound & FN (Both Models) & FN (Mouse Only) & Unbound & FP (Both Models) & FP (Mouse Only)\"\n",
    "        col_order = [\"bound\", \"bothFN\", \"mFN\", \"unbound\", \"bothFP\", \"mFP\"]\n",
    "    print(header + r' \\\\\\midrule')\n",
    "    \n",
    "    if row_order is None:\n",
    "        row_order = tfs\n",
    "    last_row = row_order[-1]\n",
    "    \n",
    "    last_repeat_type = REPEAT_TYPES[-1]\n",
    "    for repeat_type in REPEAT_TYPES:\n",
    "        print(r'\\multicolumn{7}{c}{\\textbf{' + fix_repeat_name(repeat_type) + r'}} \\\\\\midrule')\n",
    "        repeat_fracs = all_repeat_fracs[repeat_type]\n",
    "        \n",
    "        for row_key in row_order:\n",
    "            row = [repeat_fracs[row_key][col] for col in col_order]\n",
    "            row_as_str = [\"%0.1f\" % (100 * num) + r'\\%' for num in row]\n",
    "            row_as_str[-1] = r'\\textbf{' + row_as_str[-1] + r'}'\n",
    "            tf_fancy_name = tfs_latex_names[tfs.index(row_key)]\n",
    "            if row_key is not last_row:\n",
    "                print(tf_fancy_name + \" & \" + \" & \".join(row_as_str) + r' \\\\')\n",
    "            else:\n",
    "                if repeat_type == last_repeat_type:\n",
    "                    print(tf_fancy_name + \" & \" + \" & \".join(row_as_str) + r' \\\\\\bottomrule')\n",
    "                else:\n",
    "                    print(tf_fancy_name + \" & \" + \" & \".join(row_as_str) + r' \\\\\\midrule')\n",
    "\n",
    "    print(r'\\end{tabular}}{}')\n",
    "    print(r'\\captionof{table}{Percent of windows overlapping various RepeatMasker-defined ' + \\\n",
    "            'repeat elements, for different categories of genomic windows from the held-out test set. ' + \\\n",
    "            'Only RepeatMasker repeat classes with at least 500 distinct annotations within the test ' + \\\n",
    "            'set are shown. FPs: false positives. FNs: false negatives. Mouse Only: specific to' + \\\n",
    "            r'mouse-trained models. See Methods for more details on site categorization.}')\n",
    "    print(r'\\end{table*}')\n",
    "    \n",
    "\n",
    "print_full_table(all_repeat_fracs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:da2]",
   "language": "python",
   "name": "conda-env-da2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
